### DISCUSSION TITLE
2FA support

### DISCUSSION THREAD

**Discussion by jmlindemann** [rustdesk/rustdesk#3212] [Q&A] [answered]
Ability to utilize two factor authentication. 
the extra step of security to feel safe when access a remote device unattended

a good addition to the request to allow optional usage through port 443, for that extra piece of security

**Comment by lazynooblet** [comment]
I'd really like to see 2FA as a per-system configuration. I don't agree with the comment posted on #1031 saying it wouldn't make sense to have it at system level. There are some systems I want to secure access to using 2FA. Anydesk have implemented this and moving to Rustdesk feels less secure relying solely on passwords.

**Comment by rustdesk** [comment]
2FA will be ready soon since we added 2FA to our Pro for login account security lately (https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/2fa/), we have started the dev. No need to argue about account wide or per device, because we will do both. You can  choose your preference.

**Reply by lazynooblet** [reply]
This is great news, thank you.

**Comment by rustdesk** [comment, marked as answer]
Ready in nightly build now. https://github.com/rustdesk/rustdesk/releases/tag/nightly
<img width=480px src=https://github.com/rustdesk/rustdesk/assets/71636191/eb99a05e-6809-4006-bf80-e1fb8416a98e/>
<img width=480px src=https://github.com/rustdesk/rustdesk/assets/71636191/de7c81e5-fba0-48e0-bdc5-8fa5f6141f4c/>
<img width=480px src=https://github.com/rustdesk/rustdesk/assets/71636191/19ce9371-7f87-4f6b-8fdd-a0425ac1dcfd/>

**Reply by p0358** [reply]
Awesome. But enabling 2FA is kinda weird. The screen with QR code where you gotta submit the code a single time for confirmation suddenly just disappears in a few moments before I have any chance to set it up (and then is it set up or not??).

**Reply by p0358** [reply]
One more problem: it seems that the verification code simply verifies if the provided code is equal to current:
https://github.com/rustdesk/rustdesk/blob/b2290f5b4c3fd9233f6550b01c151401cdb3fed3/src/auth_2fa.rs#L90-L93

I think it should compare to at least one code before and after if not more. It is a standard practice used in a lot of software, to counteract desynchronized clocks or the user typing the code in the very last few seconds before clicking OK.

EDIT: Filed PR #7429 for that

**Comment by Sven1403** [comment]
Hello. I activated 2FA in my RustDesk but i dont get this "Two-factor authentication" window when i connect to someone? Any settings i forgot? I use RustDesk on a selfhosted server.

**Reply by xlionjuan** [reply]
Both need 1.2.4 or higher

**Reply by Sven1403** [reply]
1.2.4 is deployed in my company. Does 2FA only work when both have 2FA enabled?

**Reply by xlionjuan** [reply]
Not need

2FA will affect only on the device you've set

**Reply by Sven1403** [reply]
Well then its weird thats its not working :/

**Reply by p0358** [reply]
Are you sure you use the ***latest*** 1.2.4? Because earlier 1.2.4 nightlies did not support it yet. Also current 1.2.4 cannot connect between 1.2.3 or earlier 1.2.4, they changed something (and the 1.2.4 version number isn't final)

**Reply by Sven1403** [reply]
Ahhh i dont have newest 1.2.4 Build :) probably 2-3 weeks old. Will try newest next days. Thanks!

**Comment by CryptoZyphix** [comment]
Is there a current way with the Pro version to force 2FA for users' accounts before they can access the Address book? With Team Viewer you can enforce this before users have access to login to gain access to corp PC's in TV.
